src/ppom-controller.ts

```diff
@@ -0,0 +1,387 @@
+import {
+  BaseController,
+  BaseConfig,
+  BaseState,
+} from '@metamask/base-controller';
+import { safelyExecute } from '@metamask/controller-utils';
+import { Mutex } from 'await-semaphore';
+
+import { ppomInit, PPOM } from './ppom';
+import {
+  StorageBackend,
+  PPOMStorage,
+  PPOMFileMetadata,
+  FileInfo,
+} from './ppom-storage';
+
+const DAY = 1000 * 60 * 60 * 24;
+
+/**
+ * @type PPOMFileVersion
+ * @augments FileInfo
+ * @property filePath - Path of the file in CDN.
+ */
+type PPOMFileVersion = FileInfo & {
+  filePath: string;
+};
+
+/**
+ * @type PPOMVersionResponse - array of objects of type PPOMFileVersion
+ */
+type PPOMVersionResponse = PPOMFileVersion[];
+
+/**
+ * @type PPOMState
+ *
+ * Controller state
+ * @property lastFetched - Time when files were last updated.
+ * @property lastChainId - ChainId for which files were last updated.
+ * @property newChainId - ChainIf of currently selected network.
+ * @property versionInfo - Version information fetched from CDN.
+ * @property storageMetadata - Metadata of files storaged in storage.
+ */
+export type PPOMState = BaseState & {
+  lastFetched: number;
+  lastChainId: string;
+  newChainId: string;
+  versionInfo: PPOMVersionResponse;
+  storageMetadata: PPOMFileMetadata;
+};
+
+/**
+ * @type PPOMControllerConfig
+ *
+ * Controller configuration
+ * @property refreshInterval - Polling interval used to fetch new PPOM lists
+ */
+export type PPOMControllerConfig = BaseConfig & {
+  refreshInterval: number;
+};
+
+const PPOM_DATA_NAME = 'data';
+const PPOM_BLOB_NAME = 'blob';
+
+// TODO: replace with metamask cdn
+const PPOM_CDN_BASE_URL = 'https://storage.googleapis.com/ppom-cdn/';
+const PPOM_VERSION = 'ppom_version.json';
+const PPOM_VERSION_PATH = `${PPOM_CDN_BASE_URL}${PPOM_VERSION}`;
+
+/**
+ * PPOMController
+ * Controller responsible for managing the PPOM
+ *
+ * @property config - The controller configuration
+ * @property state - The controller state
+ * @property storage - The controller storage
+ * @property ppom - The PPOM instance
+ * @property provider - The provider used to create the PPOM instance
+ */
+export class PPOMController extends BaseController<
```

Comment:
Is it best for all new controllers to use `BaseControllerV2` to benefit from things like the controller messenger?

Reply:
I was not aware of this, let me update the PR. Thanks